### PR TITLE
Add wait until ready function

### PR DESCRIPTION
**Pull request by idonov8**
Pretty self explanatory, adding now to be used in the context of the Label Studio snippet

### CODE REVIEW

File: dagshub/data_engine/model/datasource.py [hunk 1]
```diff
@@ -343,6 +345,31 @@ def send_datapoints_to_annotation(self, datapoints: Union[List[Datapoint], List[
             webbrowser.open_new_tab(link)
         return link
 
+    def wait_until_ready(self, max_wait_time=None):
+        """
+       Blocks until the datasource preprocessing is complete
+
+       Args:
+           max_wait_time (int): Maximum time to wait in seconds
+       """
```
Comment (kbolashev) [inline dagshub/data_engine/model/datasource.py:355]: I'm not sure if having this thing indefinite by default is normal. I think we should set it to something substantial enough that a reasonable sized repo could be scanned.
10 minutes or so maybe

If the processing gets stuck at any point you don't want the users to just be stuck forever

File: dagshub/data_engine/model/datasource.py [hunk 2]
```diff
@@ -343,6 +345,31 @@ def send_datapoints_to_annotation(self, datapoints: Union[List[Datapoint], List[
...
+            self.source.get_from_dagshub()
+            if self.source.preprocessing_status == PreprocessingStatus.READY:
+                return
+
+            if self.source.preprocessing_status == PreprocessingStatus.FAILED:
+                raise RuntimeError("Datasource preprocessing failed")
+
+            if max_wait_time is not None and (time.time() - start) > max_wait_time:
+                logger.warning(f"Time limit of {max_wait_time} seconds reached before processing was completed")
+                return
```
Comment (kbolashev) [inline dagshub/data_engine/model/datasource.py]: Raise an error here that the timeout has been exceeded

File: dagshub/data_engine/model/datasource.py [hunk 3]
```diff
@@ -343,6 +345,31 @@ def send_datapoints_to_annotation(self, datapoints: Union[List[Datapoint], List[
...
 
+    def wait_until_ready(self, max_wait_time=None):
+        """
+       Blocks until the datasource preprocessing is complete
+
+       Args:
+           max_wait_time (int): Maximum time to wait in seconds
+       """
+        start = time.time()
+        logger.info("Waiting for datasource preprocessing to complete")
```
Comment (kbolashev) [inline dagshub/data_engine/model/datasource.py]: logger.info() is not seen by default.
Check out this: https://rich.readthedocs.io/en/stable/reference/status.html if you want something nicer (don't forget to use the `rich_console` from the common imports so it's not displayed in quiet mode)

**Review by kbolashev — CHANGES_REQUESTED** [review]

File: dagshub/data_engine/model/datasource.py (hunk 1 above)
Comment (kbolashev) [reply dagshub/data_engine/model/datasource.py:355]: Ask Simon what he thinks a reasonable default timeout would be

**Comment by idonov8** [comment]
Added a call to wake up the LS workspace using the a endpoint (https://github.com/DagsHub/gogs/pull/1463) to save the user some time later in the flow.

File: dagshub/data_engine/model/datasource.py [hunk 4]
```diff
@@ -343,6 +346,42 @@ def send_datapoints_to_annotation(self, datapoints: Union[List[Datapoint], List[
...
+
+            if max_wait_time is not None and (time.time() - start) > max_wait_time:
+                spinner.stop()
+                if fail_on_timeout:
+                    raise RuntimeError(f"Time limit of {max_wait_time} seconds reached before processing was completed.")
+                else:
+                    logger.warning(f"Time limit of {max_wait_time} seconds reached before processing was completed.")
+                    return
+
+            time.sleep(1)
```
Comment (kbolashev) [inline dagshub/data_engine/model/datasource.py]: The spinner is a context manager so you can do
```python
with spinner:
    do_stuff
<will stop here automatically>
```

File: dagshub/data_engine/model/datasource.py [hunk 5]
```diff
@@ -343,6 +346,42 @@ def send_datapoints_to_annotation(self, datapoints: Union[List[Datapoint], List[
...
+    def wait_until_ready(self, max_wait_time=300, fail_on_timeout=True):
+        """
+       Blocks until the datasource preprocessing is complete
+
+       Args:
+           max_wait_time (int): Maximum time to wait in seconds
+       """
+        start_workspace_url = multi_urljoin(self.source.repoApi.annotations_url, "start")
+        http_request("GET", start_workspace_url, auth=self.source.repoApi.auth)
+
```
Comment (kbolashev) [inline dagshub/data_engine/model/datasource.py]: How does this request work on the backend? Does it wait for the workspace to spin up, or returns immediately?
If it's synchronous then it's a bad idea to wait for it to complete.
Also this doesn't handle any exceptions that could occur (not talking about non-200's, but more server timeouts), so maybe wrap it in a try/catch so it doesn't block the scan accidentally

**Review by kbolashev — CHANGES_REQUESTED** [review]
A bit more changes

File: dagshub/data_engine/model/datasource.py (hunk 4 above)
Comment (idonov8) [reply dagshub/data_engine/model/datasource.py]: Cool!

File: dagshub/data_engine/model/datasource.py (hunk 5 above)
Comment (idonov8) [reply dagshub/data_engine/model/datasource.py]: Should return immediately, but maybe wait for https://github.com/DagsHub/gogs/pull/1463 to be approved to be sure

**Review by kbolashev — APPROVED** [review]

**Pull request merged**